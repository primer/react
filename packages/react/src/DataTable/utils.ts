// Utility type to generate an array with a given length. Each member of the
// array type is the length of the array
type ArrayOfLength<Length extends number, SizedArray extends Array<unknown> = []> = SizedArray['length'] extends Length
  ? SizedArray
  : ArrayOfLength<Length, [...SizedArray, SizedArray['length']]>

// A union of the valid lengths of an Array that we'll support in `ObjectPaths`
// below. This is to prevent unbounded arrays from polluting expected types
type MaxLength = ArrayOfLength<10>[number]

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ArrayIndex<A extends ReadonlyArray<any>, Keys extends number = never> = A extends readonly []
  ? Keys
  : // eslint-disable-next-line @typescript-eslint/no-unused-vars
    A extends readonly [infer _, ...infer Tail]
    ? ArrayIndex<Tail, Keys | Tail['length']>
    : Keys

// Check if the given type is within the bounds set by `MaxLength`
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ArrayWithinBounds<T> = T extends ReadonlyArray<any> & {length: infer Length}
  ? Length extends MaxLength
    ? T
    : never
  : never

// Get all valid "paths" for an object. This is useful for APIs which require
// a string input that must refer to a valid type for the data being used.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ObjectPaths<T> = T extends readonly any[] & ArrayWithinBounds<T>
  ? `${ArrayIndex<T>}` | PrefixPath<T, ArrayIndex<T>>
  : // eslint-disable-next-line @typescript-eslint/no-explicit-any
    T extends any[]
    ? never & 'Unable to determine keys of potentially boundless array'
    : T extends Date
      ? never
      : T extends object
        ? Extract<keyof T, string | number> | PrefixPath<T, Extract<keyof T, string | number>>
        : never

type PrefixPath<T, Prefix> =
  Prefix extends Extract<keyof T, number | string> ? `${Prefix}.${ObjectPaths<T[Prefix]>}` : never

// Get the value of a given path within an object
export type ObjectPathValue<ObjectType extends object, Path extends string | number> =
  ObjectType extends Record<
    string | number,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    any
  >
    ? Path extends `${infer Key}.${infer NestedPath}`
      ? ObjectPathValue<ObjectType[Key], NestedPath>
      : ObjectType[Path]
    : never
