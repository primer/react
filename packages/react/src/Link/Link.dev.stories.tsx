import Link from '.'
import type {Meta} from '@storybook/react'
import React from 'react'
import type {ComponentProps} from '../utils/types'

export default {
  title: 'Components/Link/DevOnly',
  component: Link,
} as Meta<ComponentProps<typeof Link>>

export const Inline = () => (
  <div>
    <div style={{display: 'flex', flexDirection: 'column'}} data-a11y-link-underlines="true">
      [data-a11y-link-underlines=true] (inline always underlines)
      <Link href="#">inline: undefined, underline: undefined</Link>
      <Link underline={true} href="#">
        inline: undefined, underline: true
      </Link>
      <Link underline={false} href="#">
        inline: undefined, underline: false
      </Link>
      <br />
      <Link inline={true} href="#">
        inline: true, underline: undefined
      </Link>
      <Link inline={false} href="#">
        inline: false, underline: undefined
      </Link>
      <br />
      <Link inline={true} underline={true} href="#">
        inline: true, underline: true
      </Link>
      <Link inline={true} underline={false} href="#">
        inline: true, underline: false
      </Link>
      <Link inline={false} underline={true} href="#">
        inline: false, underline: true
      </Link>
      <Link inline={false} underline={false} href="#">
        inline: false, underline: false
      </Link>
      <br />
      <Link muted={true} inline={true} href="#">
        inline: true, muted: true
      </Link>
    </div>
    <br />
    <div style={{display: 'flex', flexDirection: 'column'}} data-a11y-link-underlines="false">
      [data-a11y-link-underlines=false] (inline does nothing)
      <Link href="#">inline: undefined, underline: undefined</Link>
      <Link underline={true} href="#">
        inline: undefined, underline: true
      </Link>
      <Link underline={false} href="#">
        inline: undefined, underline: false
      </Link>
      <br />
      <Link inline={true} href="#">
        inline: true, underline: undefined
      </Link>
      <Link inline={false} href="#">
        inline: false, underline: undefined
      </Link>
      <br />
      <Link inline={true} underline={true} href="#">
        inline: true, underline: true
      </Link>
      <Link inline={true} underline={false} href="#">
        inline: true, underline: false
      </Link>
      <Link inline={false} underline={true} href="#">
        inline: false, underline: true
      </Link>
      <Link inline={false} underline={false} href="#">
        inline: false, underline: false
      </Link>
      <br />
      <Link muted={true} inline={true} href="#">
        inline: true, muted: true
      </Link>
    </div>
  </div>
)
