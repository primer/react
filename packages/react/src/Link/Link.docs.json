{
  "id": "link",
  "name": "Link",
  "status": "alpha",
  "a11yReviewed": false,
  "stories": [],
  "importPath": "@primer/react",
  "props": [
    {
      "name": "href",
      "type": "string",
      "defaultValue": "",
      "description": "URL to be used for the Link. (The `href` is passed to the underlying `<a>` element. If `as` is specified, the link behavior may need different props)."
    },
    {
      "name": "muted",
      "type": "boolean",
      "defaultValue": "false",
      "description": "Uses a less prominent shade for Link color, and the default link shade on hover."
    },
    {
      "name": "inline",
      "type": "boolean",
      "defaultValue": "false",
      "description": "Set to true for links adjacent to text, underlining them for clear visibility and improved accessibility."
    },
    {
      "name": "underline",
      "type": "boolean",
      "defaultValue": "false",
      "description": "Use `inline` instead.",
      "deprecated": true
    },
    {
      "name": "hoverColor",
      "type": "string",
      "defaultValue": "",
      "description": "Color used when hovering over the link."
    },
    {
      "name": "ref",
      "type": "React.RefObject<HTMLAnchorElement>"
    },
    {
      "name": "as",
      "type": "React.ElementType",
      "defaultValue": "\"a\""
    },
    {
      "name": "sx",
      "type": "SystemStyleObject"
    }
  ],
  "subcomponents": []
}
