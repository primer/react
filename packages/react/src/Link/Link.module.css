.Link {
  color: var(--fgColor-accent);
  text-decoration: none;

  /* Reset for button tags */
  &:is(button) {
    display: inline-block;
    padding: 0;
    font-size: inherit;
    white-space: nowrap;
    cursor: pointer;
    user-select: none;
    background-color: transparent;
    border: 0;
    appearance: none;
  }

  &:hover {
    text-decoration: underline;
  }

  /* Deprecated: but need to support backwards compatibility */
  &[data-underline='true'],
  /*
    Inline links (inside a text block), however, should have underline based on accessibility setting set in data-attribute
    Note: setting underline={false} does not override this
  */
  [data-a11y-link-underlines='true'] &[data-inline='true'] {
    text-decoration: underline;
  }

  &[data-muted='true'] {
    color: var(--fgColor-muted);

    &:hover {
      color: var(--fgColor-accent);
      text-decoration: none;
    }
  }
}
