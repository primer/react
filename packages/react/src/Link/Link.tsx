import React, {forwardRef, useEffect} from 'react'
import styled from 'styled-components'
import {system} from 'styled-system'
import {get} from '../constants'
import {useRefObjectAsForwardedRef} from '../hooks'
import type {SxProp} from '../sx'
import sx from '../sx'
import type {ComponentProps} from '../utils/types'
import type {ForwardRefComponent as PolymorphicForwardRefComponent} from '../utils/polymorphic'

type StyledLinkProps = {
  hoverColor?: string
  muted?: boolean
  /** @deprecated use `inline` to specify the type of link instead */
  underline?: boolean
  // Link inside a text block
  inline?: boolean
} & SxProp

const hoverColor = system({
  hoverColor: {
    property: 'color',
    scale: 'colors',
  },
})

const StyledLink = styled.a<StyledLinkProps>`
  color: ${props => (props.muted ? get('colors.fg.muted')(props) : get('colors.accent.fg')(props))};

  /* By default, Link does not have underline */
  text-decoration: none;

  /* You can add one by setting underline={true} */
  text-decoration: ${props => (props.underline ? 'underline' : undefined)};

  /* Inline links (inside a text block), however, should have underline based on accessibility setting set in data-attribute */
  /* Note: setting underline={false} does not override this */
  [data-a11y-link-underlines='true'] &[data-inline='true'] {
    text-decoration: underline;
  }

  &:hover {
    text-decoration: ${props => (props.muted ? 'none' : 'underline')};
    ${props => (props.hoverColor ? hoverColor : props.muted ? `color: ${get('colors.accent.fg')(props)}` : '')};
  }
  &:is(button) {
    display: inline-block;
    padding: 0;
    font-size: inherit;
    white-space: nowrap;
    cursor: pointer;
    user-select: none;
    background-color: transparent;
    border: 0;
    appearance: none;
  }
  ${sx};
`

const Link = forwardRef(({as: Component = 'a', ...props}, forwardedRef) => {
  const innerRef = React.useRef<HTMLAnchorElement>(null)
  useRefObjectAsForwardedRef(forwardedRef, innerRef)

  if (__DEV__) {
    /**
     * The Linter yells because it thinks this conditionally calls an effect,
     * but since this is a compile-time flag and not a runtime conditional
     * this is safe, and ensures the entire effect is kept out of prod builds
     * shaving precious bytes from the output, and avoiding mounting a noop effect
     */
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useEffect(() => {
      if (
        innerRef.current &&
        !(innerRef.current instanceof HTMLButtonElement) &&
        !(innerRef.current instanceof HTMLAnchorElement)
      ) {
        // eslint-disable-next-line no-console
        console.error(
          'Error: Found `Link` component that renders an inaccessible element',
          innerRef.current,
          'Please ensure `Link` always renders as <a> or <button>',
        )
      }
    }, [innerRef])
  }

  return (
    <StyledLink
      as={Component}
      data-inline={props.inline}
      {...props}
      // @ts-ignore shh
      ref={innerRef}
    />
  )
}) as PolymorphicForwardRefComponent<'a', StyledLinkProps>

Link.displayName = 'Link'

export type LinkProps = ComponentProps<typeof Link>
export default Link
