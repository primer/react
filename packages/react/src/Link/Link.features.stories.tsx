import Link from '../Link'
import type {Meta} from '@storybook/react'
import React from 'react'
import type {ComponentProps} from '../utils/types'

export default {
  title: 'Components/Link/Features',
  component: Link,
} as Meta<ComponentProps<typeof Link>>

export const Muted = () => (
  <Link href="#" muted>
    Link
  </Link>
)

export const Underline = () => (
  <Link href="#" underline>
    Link
  </Link>
)

export const Inline = () => (
  <div data-a11y-link-underlines="true">
    <Link inline={true} href="#">
      Link
    </Link>
  </div>
)
