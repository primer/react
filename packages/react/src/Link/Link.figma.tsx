import React from 'react'
import Link from './Link'
import figma from '@figma/code-connect'

/**
 * -- This file was auto-generated by `figma connect create` --
 * `props` includes a mapping from Figma properties and variants to
 * suggested values. You should update this to match the props of your
 * code component, and update the `example` function to return the
 * code example you'd like to see in Figma
 */

figma.connect(
  Link,
  'https://www.figma.com/file/GCvY3Qv8czRgZgvl1dG6lp/Primer-Web?type=design&node-id=20953-78768&mode=design&t=HqwKHI6akvFT5reK-4',
  {
    props: {
      label: figma.string('link text'),
      inline: figma.boolean('inline'),
      trailingVisual: figma.boolean('trailingVisual?', {
        true: figma.instance('trailingVisual'),
        false: undefined,
      }),
      leadingVisual: figma.boolean('leadingVisual?', {
        true: figma.instance('leadingVisual'),
        false: undefined,
      }),
      muted: figma.boolean('muted'),
    },
    example: ({label, inline, muted, leadingVisual, trailingVisual}) => (
      <Link inline={inline} muted={muted} href="#">
        {leadingVisual}
        {label}
        {trailingVisual}
      </Link>
    ),
  },
)
