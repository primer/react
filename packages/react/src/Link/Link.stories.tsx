import Link from '../Link'
import type {Meta, StoryFn} from '@storybook/react'
import React from 'react'
import type {ComponentProps} from '../utils/types'

export default {
  title: 'Components/Link',
  component: Link,
} as Meta<ComponentProps<typeof Link>>

export const Playground: StoryFn<ComponentProps<typeof Link>> = args => <Link {...args}>Link</Link>

Playground.args = {
  href: '#',
  muted: false,
  underline: false,
  inline: false,
}

export const Default = () => <Link href="#">Link</Link>
