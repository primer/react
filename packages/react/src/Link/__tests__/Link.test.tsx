import React from 'react'
import Link from '..'
import {render, behavesAsComponent, checkExports} from '../../utils/testing'
import {render as HTMLRender} from '@testing-library/react'
import axe from 'axe-core'

describe('Link', () => {
  behavesAsComponent({Component: Link})

  checkExports('Link', {
    default: Link,
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<Link href="www.github.com">GitHub</Link>)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('passes href down to link element', () => {
    expect(render(<Link href="https://github.com" />)).toMatchSnapshot()
  })

  it('respects hoverColor prop', () => {
    expect(render(<Link hoverColor="accent.fg" />)).toMatchSnapshot()
  })

  it('respects the "sx" prop', () => {
    expect(render(<Link sx={{fontStyle: 'italic'}} />)).toHaveStyleRule('font-style', 'italic')
  })

  it('applies button styles when rendering a button element', () => {
    expect(render(<Link as="button" />)).toMatchSnapshot()
  })

  it('respects the "muted" prop', () => {
    expect(render(<Link muted />)).toMatchSnapshot()
  })

  it('respects the  "sx" prop when "muted" prop is also passed', () => {
    expect(render(<Link muted sx={{color: 'fg.onEmphasis'}} />)).toMatchSnapshot()
  })

  it('logs a warning when trying to render invalid "as" prop', () => {
    const consoleSpy = jest.spyOn(global.console, 'error').mockImplementation()

    HTMLRender(<Link as="i" />)
    expect(consoleSpy).toHaveBeenCalled()

    consoleSpy.mockRestore()
  })
})
