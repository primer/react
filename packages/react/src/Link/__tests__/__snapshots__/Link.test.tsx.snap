// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link applies button styles when rendering a button element 1`] = `
.c1 {
  color: var(--fgColor-accent,var(--color-accent-fg,#0969da));
  -webkit-text-decoration: none;
  text-decoration: none;
}

[data-a11y-link-underlines='true'] .c0[data-inline='true'] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:is(button) {
  display: inline-block;
  padding: 0;
  font-size: inherit;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

<button
  className="c0 c1"
/>
`;

exports[`Link passes href down to link element 1`] = `
.c1 {
  color: var(--fgColor-accent,var(--color-accent-fg,#0969da));
  -webkit-text-decoration: none;
  text-decoration: none;
}

[data-a11y-link-underlines='true'] .c0[data-inline='true'] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:is(button) {
  display: inline-block;
  padding: 0;
  font-size: inherit;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

<a
  className="c0 c1"
  href="https://github.com"
/>
`;

exports[`Link respects hoverColor prop 1`] = `
.c1 {
  color: var(--fgColor-accent,var(--color-accent-fg,#0969da));
  -webkit-text-decoration: none;
  text-decoration: none;
}

[data-a11y-link-underlines='true'] .c0[data-inline='true'] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: var(--fgColor-accent,var(--color-accent-fg,#0969da));
}

.c1:is(button) {
  display: inline-block;
  padding: 0;
  font-size: inherit;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

<a
  className="c0 c1"
/>
`;

exports[`Link respects the  "sx" prop when "muted" prop is also passed 1`] = `
.c1 {
  color: var(--fgColor-muted,var(--color-fg-muted,#656d76));
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--fgColor-onEmphasis,var(--color-fg-on-emphasis,#ffffff));
}

[data-a11y-link-underlines='true'] .c0[data-inline='true'] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--fgColor-accent,var(--color-accent-fg,#0969da));
}

.c1:is(button) {
  display: inline-block;
  padding: 0;
  font-size: inherit;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

<a
  className="c0 c1"
  muted={true}
/>
`;

exports[`Link respects the "muted" prop 1`] = `
.c1 {
  color: var(--fgColor-muted,var(--color-fg-muted,#656d76));
  -webkit-text-decoration: none;
  text-decoration: none;
}

[data-a11y-link-underlines='true'] .c0[data-inline='true'] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--fgColor-accent,var(--color-accent-fg,#0969da));
}

.c1:is(button) {
  display: inline-block;
  padding: 0;
  font-size: inherit;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

<a
  className="c0 c1"
  muted={true}
/>
`;
